From 5f07b97078caa278fafa779029fb7f4b1fe73640 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 25 Aug 2014 07:52:39 -0400
Subject: [tools] lib/traceevent: Unregister handler when kmem plugin is unloaded

Message-id: <1408953210-26343-272-git-send-email-jolsa@redhat.com>
Patchwork-id: 88905
O-Subject: [PATCH RHEL7.1 BZ1133083 271/322] tools lib traceevent: Unregister handler when kmem plugin is unloaded
Bugzilla: 1133083
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1133083
https://bugzilla.redhat.com/show_bug.cgi?id=1133083

upstream
========
commit 11e99c55414ebade1031a0ed3b49915824c7c3ea
Author: Namhyung Kim <namhyung@kernel.org>
Date: Thu Jan 16 11:31:11 2014 +0900

description
===========
The kmem handlers should be unregistered when the plugin is unloaded
otherwise they'll try to access invalid memory.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/lib/traceevent/plugin_kmem.c b/tools/lib/traceevent/plugin_kmem.c
index 7115c80..70650ff 100644
--- a/tools/lib/traceevent/plugin_kmem.c
+++ b/tools/lib/traceevent/plugin_kmem.c
@@ -70,3 +70,25 @@ int PEVENT_PLUGIN_LOADER(struct pevent *pevent)
           call_site_handler, NULL);
  return 0;
 }
+
+void PEVENT_PLUGIN_UNLOADER(struct pevent *pevent)
+{
+ pevent_unregister_event_handler(pevent, -1, "kmem", "kfree",
+     call_site_handler, NULL);
+
+ pevent_unregister_event_handler(pevent, -1, "kmem", "kmalloc",
+     call_site_handler, NULL);
+
+ pevent_unregister_event_handler(pevent, -1, "kmem", "kmalloc_node",
+     call_site_handler, NULL);
+
+ pevent_unregister_event_handler(pevent, -1, "kmem", "kmem_cache_alloc",
+     call_site_handler, NULL);
+
+ pevent_unregister_event_handler(pevent, -1, "kmem",
+     "kmem_cache_alloc_node",
+     call_site_handler, NULL);
+
+ pevent_unregister_event_handler(pevent, -1, "kmem", "kmem_cache_free",
+     call_site_handler, NULL);
+}
-- 
1.7.1